Zend/zend_hash.c

```diff
@@ -459,12 +459,25 @@ ZEND_API void ZEND_FASTCALL _zend_hash_iterators_update(HashTable *ht, HashPosit
 	}
 }
 
-static zend_always_inline Bucket *zend_hash_find_bucket(const HashTable *ht, zend_string *key)
+/* To protect against HashDos attacks, we count collisions during the "find
+ * existing bucket with this key" phase of insertion operations. The check
+ * against MAX_COLLISIONS is only performed if *no* matching bucket is found,
+ * as that corresponds to insert operations (as opposed to updates).
+ *
+ * An important caveat of the implementation is that collisions will *not*
+ * be counted for add_new operations. find+add_new operations should be replaced
+ * with add_or_return operations.
+ */
+#define HT_MAX_COLLISIONS 1000
```

Comment:
@mfn : it is.

Ini setting is bad, because relying on the user for such a crucial, misunderstood concept, is bad.
Like we hardcoded many internal-related things, GC_ROOT_BUFFER_MAX_ENTRIES f.e.